docs/compilers/terrapin.md

```diff
@@ -0,0 +1,127 @@
+The Terrapin service will validate a DLL by going through the following workflow for every DLL in a NuPkg file:
+
+1. Execute `dotnet-build-manifest-generator` on the DLL + PDB conversion. 
+1. Setup the environment for validation:
+    1. Transfer execution to the appropriate operating system listed in the manifest. All future actions are expected to run on this machine.
+    1. Install the correct .NET Runtime on the target machine (if necessary)
+1. Download the artifacts listed in the artifact manifest file as well as the specified version of `dotnet-build-validator`. 
+1. Execute `dotnet-build-validator` providing the DLL, PDB and the directory where the artifacts were downloaded too.
+
+The Terrapin service will download the artifacts into a directory created by Terrapin. The directory will have two directories to store source files and references respectively: sources and references. The name of the artifacts in those directories will be specified in the manifest file. That will put the burden of avoiding collisions and dealing with file path fixup on the compiler team.
+
+## Artifacts Manifest File
+The goal of `dotnet-build-manifest-generator` is to provide a manifest file to the Terrapin service which lists all of the artifacts necessary for a validation to be completed as well as the environment needed to run the validation. The set of artifacts that are necessary include:
+
+1. The source files used in the compilation. 
+1. The reference DLLs used in the compilation. 
+
+At this time the expectation is Terrapin will need to recreate much of the environment used to create the binaries in order to perform provenance validation. This includes having a matching .NET Runtime version and operating system. This is because OS dependencies, like zip compression version, can impact the deterministic output of the compiler. It's possible our experiments in Terrapin will cause us to re-evaluate how these play into determinism but for now the design is to build validation on top of our existing assumptions here. 
+
+To facilitate this the artifact manifest file will include all of the necessary environment information. This will include the OS RID, the .NET Runtime version and whether a desktop or NET Core runtime was used to run the compiler.
+
+The artifacts manifest file will be a JSON file with the following format:
+
+```json
+{
+    "sources": [
+        { 
+            "uri": "https://github.com/dotnet/roslyn/blob/f5517311e37a3883e09dc6daa5069bbc4f44c882/src/Compilers/Shared/Csc.cs",
+            "destFileName": "c6fea380dac7392b945fa6c63466746f830c5cc3"
+        }
+    ],
+    "references": [
+        {
+            "name": "System.Reflection.Metadata.dll",
+            "mvid": "51ef3ce0-5a89-4325-b094-8a5970acc676",
+            "destFileName": "51ef3ce0-5a89-4325-b094-8a5970acc676",
+        }
+    ],
+    "environment": {
+        "compiler-language": "csharp",
+        "compiler-version": "3.8.0",
+        "os": "windows",
```

Comment:
do all/most of the "environment" values need to be embedded in the PDB? "os" and "os-version" aren't currently present in there.

Reply:
Eventually yes we will need this information. Particularly now that we've set it up such that the detail level of the options don't impact our ability to verify we should probably error on putting too much information here.